Rules/CommonScripts/GameStateBanners.as

```diff
@@ -0,0 +1,189 @@
+		if (driver !is null)
+		{
+			if (bannerPos != bannerDest)
+			{
+				frameTime = Maths::Min(frameTime + (getRenderDeltaTime() / maxTime), 1);
+
+				bannerPos = Vec2f_lerp(bannerStart, bannerDest, frameTime);
+			}
+
+			if (this.get_string("Animate Banner") == "win") 
+			{
+				DrawWinBanner(bannerPos, this.getTeamWon());
+				this.SetGlobalMessage("");
+			}
+			// todo: implement tth versions (alternatively remove TTH)
+			else if (this.get_string("Animate Banner") == "build" && (this.gamemode_name == "CTF" || this.gamemode_name == "SmallCTF"))
+			{
+				DrawBuildBanner(bannerPos);
+			} 
+			else if (this.get_string("Animate Banner") == "game" && (this.gamemode_name == "CTF" || this.gamemode_name == "SmallCTF"))
+			{
+				CPlayer@ p = getLocalPlayer();
+				int team = p is null ? 0 : p.getTeamNum();
+				// show flags of enemy team colour
+				if (team == 0) team = 1;
+				else team = 0;
+
+				DrawGameBanner(bannerPos, team);
+			} 
+		}
+	}
+}
+
+void onNewPlayerJoin(CRules@ this, CPlayer@ player)
+{
+	if (getLocalPlayer() !is null && getLocalPlayer().getUsername() == player.getUsername())
+	{
+		SetBanner(this);
+	}
+}
+
+void ResetBannerInfo(CRules@ this)
+{
+	this.set_string("Animate Banner", "none");
+	bannerPos = Vec2f_zero;
+	bannerDest = Vec2f_zero;
+	frameTime = 0;
+}
+
+void SetBanner(CRules@ this)
+{
+	Driver@ driver = getDriver();
+	if (driver !is null)
+	{
+		bannerDest = Vec2f(driver.getScreenWidth()/2, driver.getScreenHeight()/3);
+		bannerStart = bannerDest;
+		bannerStart.y = 0;
+		bannerPos = bannerStart;
+
+		if (this.isGameOver() && this.getTeamWon() >= 0)
+		{
+			this.set_u32("Banner End", getGameTime() + winBannerDuration);
+			this.set_string("Animate Banner", "win");
+			this.minimap = false;
+		}
+		if (this.getCurrentState() == WARMUP || this.getCurrentState() == INTERMISSION) // cringe
+		{
+			this.set_u32("Banner End", getGameTime() + buildBannerDuration);
+			this.set_string("Animate Banner", "build");
+		}
+		if (this.getCurrentState() == GAME)
+		{
+			this.set_u32("Banner End", getGameTime() + gameBannerDuration);
+			this.set_string("Animate Banner", "game");
+		}
+	}
+}
+
+void DrawWinBanner(Vec2f center, int team)
+{
+	string teamName = "Blue";
+	Vec2f offset = Vec2f_zero;
+	if (team == 1)
+	{
+		teamName = "Red";
+		offset = Vec2f(-32, -16);
+	}
+
+	Vec2f tl = center - Vec2f(160, 32);
+	Vec2f br = center + Vec2f(160, 32);
+	GUI::DrawRectangle(tl, br);
+	GUI::DrawIcon("TeamIcons.png", team, Vec2f(96, 96), center - Vec2f(96, 192) + offset, 1.0f, team);
+	GUI::DrawIcon("MenuItems.png", 31, Vec2f(32, 32), center - Vec2f(192, 32), 1.0f, team);
+	GUI::DrawIcon("MenuItems.png", 31, Vec2f(32, 32), center + Vec2f(128, -32), 1.0f, team);
+
+	GUI::SetFont("AveriaSerif-Bold_32");
+	string text = teamName + " team wins";
+	GUI::DrawTextCentered(getTranslatedString(text), center, SColor(255, 255, 255, 255));
+}
+
+void DrawBuildBanner(Vec2f center)
+{
+	string text = "Build the base!";
```

Comment:
I think "Build defences!" would be better since building a base can imply minecrafting rather than building to defend the flag.

Reply:
fixed, although it'll be the american version of defenses because kag has never used "defences"